contracts/token/ERC721/ERC721SingleRentalAgreement.sol

```diff
@@ -77,96 +79,72 @@ contract ERC721SingleRentalAgreement is Context, IERC721RentalAgreement, ERC165
             rentalStatus == RentalStatus.pending,
             "ERC721SingleRentalAgreement: rental agreement has to be pending to be updated."
         );
-        require(!rentPaid, "ERC721SingleRentalAgreement: rent already paid");
     }
 
     /// @inheritdoc IERC721RentalAgreement
-    function afterRentalStarted(address, uint256 tokenId) public override onlyErc721Contract {
-        require(block.timestamp <= expirationDate, "ERC721SingleRentalAgreement: rental agreement expired");
-        require(renter == IERC721Rental(_msgSender()).ownerOf(tokenId), "Wrong renter.");
-        require(rentalStatus == RentalStatus.pending, "ERC721SingleRentalAgreement: rental status has to be pending");
-        require(rentPaid, "ERC721SingleRentalAgreement: rent has to be paid first");
-
-        rentalStatus = RentalStatus.active;
-        startTime = block.timestamp;
-
-        // Emit an event.
-        emit RentalStatusChanged(owner, renter, tokenId, startTime, RentalStatus.pending, RentalStatus.active);
+    function afterRentalStarted(address from, uint256) public view override {
+        require(from == address(this));
     }
 
-    /// Enables the renter to pay the rent.
-    /// The rent has to be paid in order to start the rent agreement.
-    function payRent() public payable {
+    /// Enables the caller to pay and start the rent, and become the renter.
+    function payAndStartRent() public payable {
+        require(rentalStatus == RentalStatus.pending, "ERC721SingleRentalAgreement: rental status has to be pending");
         require(block.timestamp <= expirationDate, "ERC721SingleRentalAgreement: rental agreement expired");
-        require(!rentPaid, "ERC721SingleRentalAgreement: rent already paid");
-        require(_msgSender() == renter, "ERC721SingleRentalAgreement: renter has to pay the rental fees");
         require(msg.value == rentalFees, "ERC721SingleRentalAgreement: wrong rental fees amount");
 
-        rentPaid = true;
-        balances[owner] += msg.value;
+        // Start rental.
+        rentalStatus = RentalStatus.active;
+        ownerBalance += msg.value;
+        startTime = uint40(block.timestamp);
+
+        // Accept rental agreement between owner and msg sender.
+        renter = _msgSender();
+        erc721Contract.acceptRentalAgreement(renter, tokenId);
 
         // Emit event.
         emit RentPayment(owner, renter, msg.value);
+        emit RentalStatusChanged(owner, renter, tokenId, startTime, RentalStatus.pending, RentalStatus.active);
     }
 
     /// @inheritdoc IERC721RentalAgreement
-    function afterRentalStopped(address from, uint256 tokenId) public override onlyErc721Contract {
+    function afterRentalStopped(address from, uint256) public override onlyErc721Contract {
         require(rentalStatus == RentalStatus.active, "ERC721SingleRentalAgreement: rental status has to be active");
```

Comment:
You have to check the tokenId as well, otherwise I can set that contract on another token and stop it from there

Reply:
that's right, I added it, good catch!